### PR TITLE
Enable --cfg=job --resolve with ${hydra:...} resolver

### LINKED ISSUE #1681: [Bug] HydraConfig is not set while resolving config with --cfg job --resolve
```
$ python my_app.py  '+a=${hydra:runtime.cwd}/data/MNIST' --cfg job --resolve
...
omegaconf.errors.InterpolationResolutionError: ValueError raised while resolving interpolation: HydraConfig was not set
```

### PR DESCRIPTION
**Pull request by Jasha10**
This PR closes #1681, so that resolved references to `HydraConfig` can appear
in the output from the `--cfg=job` flag.  It is now possible to call
    ```python my_app.py --cfg job --resolve```
when the user config contains the `${hydra:...}` resolver.

To implement this, several changes are made to the
`hydra._internal.hydra.Hydra` class:
  - In the `Hydra._get_cfg` method, we now call
    `HydraConfig.instance().set_config(cfg)` on the composed `cfg` object,
    enabling the `${hydra:...}` resolver.
  - Previously, the `Hydra._get_cfg` method would delete the `cfg["hydra"]`
    package immediately. We now delay the deletion until after
    `OmegaConf.resolve` can be called. This is done by factoring out a
    `Hydra.get_sanitized_cfg` method.
  - Since the `HydraConfig.set_config` method sets a `'readonly'` flag, it is
    now necessary to use a `read_write`/`flag_override` context manager when
    calling `OmegaConf.resolve` and when deleting the `cfg["hydra"]` package.
  - We also call `HydraConfig.instance().set_config(cfg)` in the
    `Hydra.app_help` method. This allows for
      ```python my_app.py --help --resolve```
    to work even if the help template references the user's `$CONFIG` and the
    `${hydra:...}` resolver appears in that user config.

### CODE REVIEW

File: hydra/_internal/hydra.py [hunk 1]
```diff
@@ -156,16 +157,19 @@ def _get_cfg(
...
         with_log_configuration: bool,
     ) -> DictConfig:
-        assert cfg_type in ["job", "hydra", "all"]
         cfg = self.compose_config(
             config_name=config_name,
             overrides=overrides,
             run_mode=RunMode.RUN,
             with_log_configuration=with_log_configuration,
         )
+        HydraConfig.instance().set_config(cfg)
```
Comment (omry) [inline hydra/_internal/hydra.py:168]: I would prefer not to introduce a side effect to a getter function.
Can the caller do this instead?

File: hydra/_internal/hydra.py [hunk 2]
```diff
@@ -204,7 +207,9 @@ def show_cfg(
             if package is not None:
                 print(f"# @package {package}")
             if resolve:
-                OmegaConf.resolve(ret)
+                with read_write(cfg.hydra):
+                    OmegaConf.resolve(ret)
```
Comment (omry) [inline hydra/_internal/hydra.py:211]: This seems unrelated. What is the motivation for this? (Is there a missing test?)

**Review by omry — COMMENTED** [review]
nice. See comment.

File: hydra/_internal/hydra.py (hunk 2 above)
Comment (Jasha10) [reply hydra/_internal/hydra.py:198]: The `HydraConfig.set_config` method sets a `readonly` flag on `cfg.hydra`, so `OmegaConf.resolve` fails unless the flag is turned off.

File: hydra/_internal/hydra.py (hunk 1 above)
Comment (Jasha10) [reply hydra/_internal/hydra.py]: I inlined the `_get_cfg` method in https://github.com/facebookresearch/hydra/pull/1696/commits/6828fcc003618a8a4dafd5a01ed3653fdd93655d.

File: hydra/_internal/hydra.py [hunk 3]
```diff
@@ -147,25 +148,13 @@ def get_sanitized_hydra_cfg(src_cfg: DictConfig) -> DictConfig:
             for key in list(cfg.keys()):
                 if key != "hydra":
                     del cfg[key]
-        with open_dict(cfg.hydra):
+        with flag_override(cfg.hydra, ["struct", "readonly"], [False, False]):
```
Comment (omry) [inline hydra/_internal/hydra.py:151]: This should work:
```suggestion
        with flag_override(cfg.hydra, ["struct", "readonly"], False):
```

File: hydra/_internal/hydra.py (hunk 2 above)
Comment (omry) [reply hydra/_internal/hydra.py:198]: We could unset the readonly flag in show_cfg and show_help.

File: tests/test_hydra.py [hunk 4]
```diff
@@ -481,11 +483,23 @@ def test_cfg_with_package(
...
 )
-def test_cfg_resolve_interpolation(tmpdir: Path, resolve: bool, expected: str) -> None:
+def test_cfg_resolve_interpolation(
+    tmpdir: Path, script: str, resolve: bool, expected: str
+) -> None:
     cmd = [
-        "tests/test_apps/simple_interpolation/my_app.py",
+        script,
         "hydra.run.dir=" + str(tmpdir),
         "--cfg=job",
```
Comment (omry) [inline tests/test_hydra.py:504]:
```suggestion
        script, f"hydra.run.dir={tmpdir}", "--cfg=job"
```

File: news/1681.bugfix [hunk 5]
```diff
@@ -0,0 +1 @@
+Fix the --cfg=job --resolve flags so that the ${hydra:...} resolver now prints properly
```
Comment (omry) [inline news/1681.bugfix:1]: This fixes --help as well?

File: tests/test_apps/interpolation_to_hydra_config/my_app.py [hunk 6]
```diff
@@ -0,0 +1,13 @@
+# Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved
+from omegaconf import DictConfig
+
+import hydra
+
+
+@hydra.main(config_path=".", config_name="config")
+def my_app(cfg: DictConfig) -> None:
+    pass
```
Comment (omry) [inline tests/test_apps/interpolation_to_hydra_config/my_app.py:9]: Can you reuse simple_app here?
You can place the config next to it or somewhere else and override the used config like:
```
python tests/test_apps/simple_app/my_app.py  --config-name=config --config-dir=tests/test_apps/app_with_cfg
```

File: tests/test_hydra.py [hunk 7]
```diff
@@ -617,6 +631,20 @@ def test_sweep_complex_defaults(
...
+            ["hydra.help.template=$CONFIG", "db.user=${hydra:job.name}"],
+            dedent(
+                """\
+                db:
+                  driver: mysql
+                  user: my_app
+                  password: secret
+                """
+            ),
+            id="overriding_help_template:$CONFIG,resolve_interp_to_hydra_config",
```
Comment (omry) [inline tests/test_hydra.py:646]: generally speaking, avoid testing with the tutorial apps except in tests/test_examples. (I know the code is not being consistent about it).

File: hydra/_internal/hydra.py (hunk 2 above)
Comment (Jasha10) [reply hydra/_internal/hydra.py]: Done in https://github.com/facebookresearch/hydra/pull/1696/commits/57452d5b9e71485cb3e92d6525f2d71a2cb3abc5. I only needed to unset the flag in `show_cfg`. In `show_help`, a readonly flag is no problem because we are using `OmegaConf.to_yaml(..., resolve=...)` instead of `OmegaConf.resolve(...)`.

File: hydra/_internal/hydra.py [hunk 8]
```diff
@@ -141,37 +142,28 @@ def multirun(
...
 
     @staticmethod
-    def get_sanitized_hydra_cfg(src_cfg: DictConfig) -> DictConfig:
-        cfg = copy.deepcopy(src_cfg)
+    def sanitize_hydra_cfg(cfg: DictConfig) -> None:
+        """
+        Mutate the input cfg:
+          - delete all toplevel keys except "hydra".
+          - delete `hydra.help` and `hydra.hydra_help`.
+        """
```
Comment (Jasha10) [inline hydra/_internal/hydra.py:150]: https://github.com/facebookresearch/hydra/pull/1696/commits/57452d5b9e71485cb3e92d6525f2d71a2cb3abc5 is a refactoring:
  - `get_sanitized_hydra_cfg` -> `sanitize_hydra_cfg`
  -  `get_sanitized_cfg` -> `sanitize_cfg`


Before, these methods were making a deepcopy in some branches, and not making a deepcopy in other branches.
Now, they never make a deepcopy (i.e. they mutate their input directly). This refactoring should resolve the LGTM alert that was posted in the PR's main thread..

File: news/1681.bugfix (hunk 5 above)
Comment (Jasha10) [reply news/1681.bugfix]: Yes. I updated the news fragment in https://github.com/facebookresearch/hydra/pull/1696/commits/751653f665956141f43453eb30814cd72513d694.

File: hydra/_internal/hydra.py (hunk 8 above)
Comment (omry) [reply hydra/_internal/hydra.py:150]: Looks like get_sanitized_hydra_cfg was always copying and get_sanitized_cfg was composing (so no need to copy).
Not sure what alerts you are referring to.
```python
def get_sanitized_hydra_cfg(src_cfg: DictConfig) -> DictConfig:
        cfg = copy.deepcopy(src_cfg)
```

Comment (omry) [reply hydra/_internal/hydra.py:150]: Looking at the warning, the refactoring sounds like way too big a hammer.

image

Comment (Jasha10) [reply hydra/_internal/hydra.py]: Fair enough. I reverted the refactoring in https://github.com/facebookresearch/hydra/pull/1696/commits/f0fa6cdf104fb3a93a9e9a220f040ea02e8892b3 and then did a simpler fix in https://github.com/facebookresearch/hydra/pull/1696/commits/06258d203ae51cba45659b6e4eb3bb8b7a8e71fd (calling `get_sanitized_cfg` earlier).

File: tests/test_hydra.py (hunk 7 above)
Comment (Jasha10) [reply tests/test_hydra.py]: Made the change in https://github.com/facebookresearch/hydra/pull/1696/commits/27052cf96c4261051b4eea432e34ccbabbb1e22e.

File: tests/test_apps/interpolation_to_hydra_config/my_app.py (hunk 6 above)
Comment (Jasha10) [reply tests/test_apps/interpolation_to_hydra_config/my_app.py]: Done in https://github.com/facebookresearch/hydra/pull/1696/commits/1d7bdd20fc4928af6b86d5c9b0f20625afbc9497.

**Review by omry — APPROVED** [review]
Looking good.
